compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/extended/VariableAssignmentChecker.kt

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.extended
+
+
+import org.jetbrains.kotlin.contracts.description.EventOccurrencesRange
+import org.jetbrains.kotlin.fir.FirFakeSourceElement
+import org.jetbrains.kotlin.fir.analysis.cfa.*
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.checkers.declaration.FirBasicDeclarationChecker
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.declarations.FirDeclaration
+import org.jetbrains.kotlin.fir.declarations.FirFunction
+import org.jetbrains.kotlin.fir.psi
+import org.jetbrains.kotlin.fir.references.FirResolvedNamedReference
+import org.jetbrains.kotlin.fir.resolve.dfa.cfg.*
+import org.jetbrains.kotlin.fir.resolve.dfa.controlFlowGraph
+import org.jetbrains.kotlin.fir.symbols.impl.FirPropertySymbol
+import org.jetbrains.kotlin.fir.toFirPsiSourceElement
+import org.jetbrains.kotlin.psi.KtProperty
+
+object VariableAssignmentChecker : FirBasicDeclarationChecker() {
+    override fun check(declaration: FirDeclaration, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (declaration !is FirFunction<*>) return
+
+        val graph = declaration.controlFlowGraphReference.controlFlowGraph ?: return
+        val unprocessedProperties = mutableSetOf<FirPropertySymbol>()
+        val propertiesCharacteristics = mutableMapOf<FirPropertySymbol, AssignmentsCount>()
+
+        val visitor = ControlFlowGraphDeclarationVisitor(unprocessedProperties, propertiesCharacteristics)
+        visitor.analyze(graph, reporter)
+
+        for (property in unprocessedProperties) {
+            if (property.fir.source is FirFakeSourceElement<*>) continue
+            if (property.callableId.callableName.asString() == "<destruct>") continue
+            propertiesCharacteristics[property] = AssignmentsCount.ZERO
+        }
+
+        for ((symbol, value) in propertiesCharacteristics) {
+            val source = symbol.valOrVarOrJustKeywordSource
+            when (value) {
+                AssignmentsCount.ZERO -> {
+                    reporter.report(source, FirErrors.UNINITIALIZED_VARIABLE_EXTENDED)
```

Comment:
I think this duplicated standard `UNINITIALIZED_VARIABLE`, so we should either unite both checkers or, at least, do not report this diagnostic here

Reply:
This is true interesting moment. This code reports `UNINITIALIZED_VARIABLE_EXTENDED` even variable not used. The standert checker reports `UNINITIALIZED_VARIABLE` only if value in variable reads somewhere. So I could remove this report or leave it. I'm not know which way is right